libqtile/core/manager.py

```diff
@@ -527,25 +527,40 @@ def current_group(self):
     def current_window(self):
         return self.current_screen.group.current_window
 
-    def add_strut(self, strut):
+    def add_strut(self, strut, screen):
         from libqtile.bar import Bar, Gap
 
+        if any(e is None for e in [screen, screen.x, screen.y, screen.height, screen.width]):
+            logger.error("Missing screen information")
+            return
+
+        x_screen_dimensions = self.root.get_geometry()
+        gaps = [
+            screen.x,
+            x_screen_dimensions.width - screen.x - screen.width,
+            screen.y,
+            x_screen_dimensions.height - screen.y - screen.height,
+        ]
```

Comment:
As you've moved some of the calculations into `class Static`, maybe calculate `gaps` there too and pass `screen` and `gaps` to `Qtile.add_strut`. This avoid doing `x_screen_dimensions = self.root.get_geometry()` in two places.

It also appears that if `qtile.find_screen` returns `None` then line 533 will get an `AttributeError` when trying to get `screen.x`. Maybe upon return of that function only continue the calculations and call to `add_strut` if `screen is not None`.

A nice side benefit of this is that these calculations, which are specific to X's strut hints, are separated from the more general "use `gaps` to update this screen's `Gap`s" which in theory could be used as-is with any backend (just thinking long-term with the come of Wayland).

Reply:
Moving the gaps calculations into `class Static`, would you suggest storing the list in a new attribute `self.gaps` for instances of Static? We will need the values again to remove struts and this call is not done from within `update_strut`.

Is there a way to stop evaluating the list as soon as a `None` value is found?

Reply:
The `self.strut` attribute already stores the window property from `_NET_WM_STRUT_PARTIAL` so maybe it makes sense to store the calculated version there instead of the window property. We'd need to re-process it when this attribute is used for `remove_strut` anyway so this way it's already done. I think that makes sense?

> Is there a way to stop evaluating the list as soon as a None value is found?

Which bit are you referring to?

Reply:
Yes, storing it probably would make sense. Do we then actually need the pure values of `_NET_WM_STRUT_PARTIAL` in `self.strut`? I think I'd like to store just one global attribute, something like `self.reserved_space`, which we generate in `update_strut` from the local `strut` and `gaps`? This way we'd only need to pass two arguments.

Reply:
> Is there a way to stop evaluating the list as soon as a None value is found?
I'm referring to the `AttributeError`. I am used to such evaluations being lazy and therefore not checking as far as `screen.x` if `screen` itself is None.

Reply:
> Do we then actually need the pure values of _NET_WM_STRUT_PARTIAL in self.strut?

I don't think so. I think your approach with saving gaps in `self.reserved_space`\* and removing `self.strut` is a good idea. I guess that and `self.screen` is sufficient for removal, right?

For the `AttributeError` -- I think dropping `screen` from that list altogether would be fine. Instead you could check that the return value of `qtile.find_screen` is not `None` and only call `qtile.add_strut` if it is alright. This way, `qtile.add_strut` can just always except a Screen object.

\* Calling it reserved_space is actually really nice, maybe we should rename the methods that refer to "strut" to something similar, which would probably make things more explicit and clear. For instance `Qtile.add_strut` could definitely be improved by renaming it to `Qtile.reserve_space` or something. Likewise `Qtile.unreserve_space` or something to remove them. Anyway this is unnecessary so don't feel like you need to make those changes!

Reply:
> This way, `qtile.add_strut` can just always except a Screen object.

Moving the `gaps` logic to `class Static` introduces the need to already check the whole list `update_strut`.

Reply:
Ah sorry, bad wording, that's what I meant :)